### PR TITLE
feat(Interfaces): Extract pure interfaces, abstract classes, and primitives into separate build artifact

### PR DESCRIPTION
**Pull request by matthargett**
Create ReactNative.Interfaces and move most pure interfaces, abstract… classes, related Exception and EventArgs there. Extract IReactContext, IReactChoreographer, and parameterize. Add primitive properties to IMutableFrameEventArgs to allow it to replace concrete FrameEventArgs type in more places. TODO: Discuss path for IViewManager, NativeModuleBase, and ReactContextNativeModuleBase, which are holding back other pure interfaces such as IReactPackage. Adjust CLI template and npm packaging to include ReactNative.Interfaces.

### CODE REVIEW

**Comment by matthargett** [comment]
Note: I can pull in the CircleCI yml from the netstandard branch when we’re ready. I was asked to disable it in that other PR because the CircleCI enablement was viral to all PRs.

File: ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace ReactNative.Modules.Core
+{
+    public interface IMutableFrameEventArgs
```
Comment (rozele) [inline ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs]: >IMutableFrameEventArgs [](start = 21, length = 22)

This interface is an implementation detail. We shouldn't move this one.

File: ReactWindows/ReactNative.Interfaces/Bridge/IReactContext.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+namespace ReactNative.Bridge
+{
+    /// <inheritdoc />
+    /// <summary>
+    /// Abstract context wrapper for the React instance to manage
+    /// lifecycle events.
+    /// </summary>
+    public interface IReactContext : IAsyncDisposable
+    {
```
Comment (rozele) [inline ReactWindows/ReactNative.Interfaces/Bridge/IReactContext.cs]: We'll want to minimize what we expose to modules. I'm thinking the interface only needs to expose
- Get*Module<T>
- Add/Remove*Listener
- IsOn*Thread
- AssertOn*Thread
- RunOn*Thread
- HandleException

File: ReactWindows/ReactNative.Interfaces/Bridge/IGenericDelegate.cs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 namespace ReactNative.Bridge
 {
-    interface IGenericDelegate
+    public interface IGenericDelegate
```
Comment (rozele) [inline ReactWindows/ReactNative.Interfaces/Bridge/IGenericDelegate.cs]: >IGenericDelegate [](start = 21, length = 16)

This is an implementation detail of our INativeModule base class implementation. We don't need this interface.

File: ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs (hunk 1 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs]: It looks like I forgot to include the extraction of IReactChoreographer in this first commit. ReactChoreographer has several public events (IdleCallback, etc), and the interface would seemingly need to expose that type as EventHandler<IMutableFrameEventArgs>.

File: ReactWindows/ReactNative.Interfaces/Bridge/IReactContext.cs (hunk 2 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Interfaces/Bridge/IReactContext.cs]: #fixed

Comment (matthargett) [reply ReactWindows/ReactNative.Interfaces/Bridge/IReactContext.cs]: We should do an inventory of native module add-ons that support RN Windows and make sure they can work with this reduced surface area. linear-gradient, video, ..?

File: ReactWindows/ReactNative.Interfaces/Bridge/IGenericDelegate.cs (hunk 3 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Interfaces/Bridge/IGenericDelegate.cs]: #fixed

**Comment by matthargett** [comment]
It looks like Any CPU got added in due to the inclusion of a netstandard library. It doesn't incorrectly add AnyCPU for the other targets, which is good, but I wonder how this affects DX (if at all). @rozele thoughts?

Also, how do I test my changes to local-cli? I thought we had documented them somewhere, but couldn't find them in the Contribution markdown file. If someone can spell it out for me, I'll update the Contribution file with the test plan people should follow after updating local-cli and/or package.json.

File: ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs (hunk 1 above)
Comment (rozele) [reply ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs]: Let's just make the event types for `IReactChoreographer`:
```
event EventHandler<FrameEventArgs>
```

Comment (rozele) [reply ReactWindows/ReactNative.Interfaces/Modules/Core/IMutableFrameEventArgs.cs]: Or create an `IFrameEventArgs`.

**Comment by rozele** [comment]
Hmm - maybe we should delete the Any CPU platform target for now?

---
In reply to: [379422923](https://github.com/Microsoft/react-native-windows/pull/1750#issuecomment-379422923) [](ancestors = 379422923)

**Comment by rozele** [comment]
> namespace ReactNative.UIManager

We should be able to move IViewManager now as well.

---
Refers to: ReactWindows/ReactNative.Interfaces/UIManager/Dimensions.cs:6 in bd68ab5. [](commit_id = bd68ab54240766424259018d1ce20f2c2f8f421d, deletion_comment = False)

**Comment by matthargett** [comment]
I'll rebase this once the IReactContext and IncrementJSCalls PRs are merged, then move over IViewManager (and re-evaluate what else that frees up for moving over).

**Pull request closed without merging**